JPA 3단계 - 질문 삭제하기 리팩토링 PR 요청 드립니다.

ssonsh opened the pull request:
이번 미션은 기능 요구사항을 하나씩 따라가면서 기능목록을 작성하고 진행하였더니
마지막 부분에서 대대적인 리팩토링이 일어나게 되었습니다. ㅠ

건별로 기능을 개발하고 나서 연관관계에 따른 처리가 이뤄지다보니 후반부 커밋이 마음에 들지 않은 상태로 올라가게 된 것 같습니다.

이번 미션을 진행하면서 피드백 주셨던 사항들에 대해 다시한번 짚고 넘어갈 수 있는 시간이 되었습니다. :)

바쁘신와중에도 정성껏 피드백 주셔서 감사합니다. 마지막 미션이네요! 👍  리뷰어님도 화이팅입니다!

ssonsh committed:
refactor : getter, setter에 대한 정리

ssonsh committed:
ref거ctor : 의미 없는 테스트코드 제

ssonsh committed:
docs(README) : 3단계 미션을 위한 1차 기능목록 도출

ssonsh committed:
feat(Question) : Question 데이터를 삭제하기 위한 api를 제공한다.

- 삭제를 위한 toDeleted() 메소드를 제공한다.

- 실제 데이터 삭제가 아니라 deleted 값을 true로 변경해준다.

ssonsh committed:
test(QuestionTest) : 질문 삭제 도메인 메소드 테스트 코드 작성

- 기존 QuestionTest 를 Repository Test 로 변경

- Question 도메인을 위한 테스트 코드 추가

ssonsh committed:
docs(README) : 질문 데이터를 삭제 시 삭제 상태로 변경하는 기능개발에 대한 목록 업데이트

ssonsh committed:
feat(Question) : 질문 데이터 삭제 시 질문 작성자와 로그인한 사용자가 같은지 유효성 검증을 진행한다.

- 제공되는 toDeleted() 삭제 API에 로그인한 사용자 정보를 매개변수로 추가한다.

- 이후 toDeleted() 내부에서 작성된 Writer와 전달된 loginUser가 일치하는지 여부를 판단한다.

- 일치하지 않을 경우 CannotDeleteException이 발생한다.

ssonsh committed:
test(QuestionTest) : 질문 데이터 삭제 사용자에 대한 유효성 검증 테스트 코드 작성

- 작성자가 같을 경우 정상적으로 삭제 된다.

- 일치하지 않을 경우 CannotDeleteException이 발생한다.

ssonsh committed:
docs(README) : 로그인한 사용자와 질문한 사람이 같은 경우에 삭제 할 수 있는 기능 개발에 대한 목록 업데이트

ssonsh committed:
feat(Question) : 질문 삭제 시 작성된 답변이 있는 경우 질문 작성자와 답변 작성자가 동일해야 삭제 가능하다.

- 질문의 답변 도메인에 대한 작성자 유효성 검증 처리 추가

- 질문자와 답변자가 다른 경우 CannotDeleteException 발생에 대한 테스트 코드 작성

ssonsh committed:
docs(README) : 질문 작성자와 답변 작성자가 같은 경우에만 질문을 삭제할 수 있는 기능 개발에 대한 목록 업데이트

ssonsh committed:
feat(Answer) : 질문 삭제 시 답변도 삭제되어야 하며 답변 삭제 또한 삭제 상태를 변경한다.

- 삭제를 위한 toDelete() 메소드 제공

ssonsh committed:
docs(README) : 질문 삭제 시 답변 또한 삭제되어야 한다는 기능목록에 대한 업데이트

ssonsh committed:
refactor(Answer) : 삭제 기능에 대한 리팩토링 진행

- validate 로직을 해당 도메인이 처리할 수 있도록 이동

- 불필요 메소드 제거

ssonsh committed:
refactor(Answers) : Answer 도메인에 대한 일급 컬랙션 구성 및 테스트 코드 작성

- Question과 Answer의 연관관계는 일대다 관계로써 다 영역의 Answer 도메인의 일급 컬렉션을 구성한다.

- Question 에서는 이 Answers 일급 컬렉션가 연관관계를 정의한다.

ssonsh committed:
refactor(Question, DeleteHistory) : Question 도메인과 Answer 도메인간의 연관관계 처리 및 DeleteHistory 처리

- Question 도메인은 Answer 도메인가 일대다 관계를 형성한다.

- Question 도메인에서는 Answer를 추가할 수 있으며 이와 관련하여 연관관계 편의 메소드를 정의하여 관리된다.

- Question의 삭제 결과로 DeleteHistory가 생성된다.

ssonsh committed:
docs(README) : 연관관계 리팩토링 및 추가 사항에 대해 기능목록을 추가하고 처리한다.

ssonsh committed:
refactor(User) : 미사옹 메소드 제거 및 public 메소드 테스트 코드 작성

ssonsh committed:
refactor(User) : 미사옹 메소드 제거 및 public 메소드 테스트 코드 작성

- 미사용 import 처리

ssonsh committed:
refactor(Question) : public으로 제공되지 않아도 될 메소드에 대한 접근지시자 변경

- public -> private

ssonsh committed:
refactor(DeleteHistory) : public 생성자에 대해 접근 지시자 private로 변경

ssonsh committed:
refactor(Answer) : 테스트만을 위해 public으로 열려있던 접근지시자 private로 변경하며 public method에 대해 테스트 코드 작성

- answer 삭제는 본인이 작성한 answer만 가능하다.

- 본인이 아닌 경우 CanntDeleteException 발생에 대한 테스트 코드 작성

ssonsh committed:
style : 코드 스타일 포매팅

wooobo commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -56,48 +58,45 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public void delete() {
-        this.deleted = true;
+    private void changeDeleted(boolean deleted) {
```
`deleted`로도 충분한 의미가 전달될것 같은데 어떻게 생각하시나요?

`change`를 붙이면 중복의 의미일것 같은 개인적인 생각입니다 🤔

ssonsh replied:
"넘어오는 인자에 따라 deleted 속성 값을 바꾼다" 라는 개념으로 changeDeleted 라는 네이밍을 사용하게 되었습니다!
하지만 말씀주신대로 change를 붙이지 않더라도 충분한 의미가 전달된다 라는점 충분히 공감합니다!
:)

ssonsh replied:
changeDeleted -> deleted 로 변경하였습니다!

wooobo commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,42 @@
+@Embeddable
+public class Answers {
+    @OneToMany(mappedBy = "question")
+    private final List<Answer> elements;
```
`answers` 도 있을것 같은데 `elements`로 네이밍 하신 이유가 궁금합니다!

ssonsh replied:
이 부분에 대해선 함께 얘기해보면 좋을 것 같아요! 

실제 Answers 라는 일급 컬랙션으로 지정해두었기 때문에 그 내부에는 요소, 값 이라는 개념으로
elements, values 와 같은 네이밍을 써도 좋을 것 같다는 의견으로 작성을 했습니다.

외부로 감춰진 녀석임으로 Answers 일급 컬렉션을 구성하는 요소들이라는 관점으로 elements를 사용한 부분이라
리뷰어님께서는 어떤 방향성을 선호하시나요? ㅎ

실제 현업에서는 values 라는 네이밍을 주로 사용하고 있긴 합니다!

wooobo replied:
저도 참 애매하다고 생각하는 부분인데요!

저는 `answers` 를 선호 하는것 같습니다ㅎ

다른분들의 의견도 궁금하시면 [# wootecam-pro-qna 채널](https://nextstep-members.slack.com/archives/C0253RP321J)
를 활용해보세요~! 👍

wooobo commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public List<DeleteHistory> deleteAll(User owner) throws CannotDeleteException {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
```
`DeleteHistories` 일급컬렉션을 잊으신것 같아요! 
아니면 이렇게 만드신 이유가 있으신가요?

도움되는 블로그 첨부해드립니다!
[불변 객체로 만드세요](https://jessyt.tistory.com/148)
[[엘레강트 오브젝트] 3-4장 충성스러우면서 불변이거나, 아니면](https://jessyt.tistory.com/157?category=967484)

ssonsh replied:
아 실제로 DeleteHistories 에 대한 일급 컬랙션이 존재하고 있으나
Answer의 deleteAll 의 결과로 List 컬랙션을 반환하고 있는데요
이 반환된 List Collection은 Question 도메인에서 받은 다음 DeleteHistories 일급 컬랙션으로 최종 생성을 하고 있습니다.

그 이유는 Question 도메인에서 Question 자체를 삭제함에 따른 DeleteHistory 엔티티와 Answers 를 삭제함에 따른 DeleteHistory들을 모아서 한번에 만들어주려는 목적으로 하였습니다.

이 부분 한번 리팩토링 해보겠습니다! 리팩토링의 방향성에 대한 피드백도 주시면 감사하겠습니다!

ssonsh replied:
리팩토링의 방향성을 Answers의 deletedAll 메소드 결과로 DeleteHistories 일급 컬랙션을 반환하고
이를 전달받은 Question 에서는 일급 컬랙션을 merge하여 최종 반환을 하는 식으로 해보았습니다.

이렇게 해보다보니
Question 입장에서 만들어내는 DeleteHistories 객체가 하나 생성이되고
Answers 의 deleteAll 에서 만들어지는 DeleteHistories 객체가 하나더 생성되게 되었습니다.

제가 방향을 잘못잡았을 수 있을 것 같은데 피드백 부탁드리겠습니다. !

wooobo commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,24 +29,7 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.toDeleted(loginUser);
```
리팩토링 💯

wooobo reviewed (changes requested):
스텝별로 커밋을 잘 나누어 주셔서 코드를 이해하고 리뷰하기에 좋았습니다  감사합니다 👍

깜끔하게 잘구현해주셔셔 바로 `merge` 하고싶지만,
궁금한 부분이 있어서 몇가지 질문 남겨드렸는데 확인 부탁드려요!

ssonsh committed:
refactor : DeleteHistories 일급 컬랙션 활용으로 변경

ssonsh committed:
refactor : 메소드 네이밍 변경

- 충분히 인지가능한 수준의 네이밍이라면 굳이 change와 같은 동사를 붙이지 않아도 될 것 같다.

- 오히려 중복된 의미를 내포하게 되는 것 같다

wooobo commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -24,12 +24,12 @@ public static Answers createEmpty() {
             deleteHistories.add(answer.toDeleted(owner));
         }
-        return deleteHistories;
+        return DeleteHistories.of(deleteHistories);
```
>리팩토링의 방향성을 Answers의 deletedAll 메소드 결과로 DeleteHistories 일급 컬랙션을 반환하고
이를 전달받은 Question 에서는 일급 컬랙션을 merge하여 최종 반환을 하는 식으로 해보았습니다.
이렇게 해보다보니
Question 입장에서 만들어내는 DeleteHistories 객체가 하나 생성이되고
Answers 의 deleteAll 에서 만들어지는 DeleteHistories 객체가 하나더 생성되게 되었습니다.
제가 방향을 잘못잡았을 수 있을 것 같은데 피드백 부탁드리겠습니다. !

=> 충분히 좋다고 생각합니다!

`List<DeleteHistory>` 으로 리턴 할 경우 외부에서 데이터를 변경가능한 상태이기 때문에 프로세스의 안정성이 떨어진다고 생각했기때문에 추가 피드백 드린 부분이었습니다 :)
하지만, 어쩔수 없는경우라면  `Collections.unmodifiableList` 을 활용해볼수도 있겠죠!

wooobo replied:
# 일급콜렉션 
> 비지니스에 종속적인 자료구조
Collection의 불변성을 보장
상태와 행위를 한 곳에서 관리
이름이 있는 컬렉션


좀 더 일급콜렉션을 활용할수 있지 않을까 생각해보았어요!
저의 관점을 좀더 공유 시켜드리기 위해 코드 첨부해드립니다.
```
DeleteHistories  deleteHistories = DeleteHistories::empty()

for (Answer answer : this.elements) {
    deleteHistories.add(answer.toDeleted(owner));
}
return deleteHistories;
```

ssonsh replied:
정성스런 피드백 감사합니다! :)

wooobo reviewed (approved):
`JPA` 미션을  빨리 완주해주셨네요! 💯 

진행하느라 고생많으셨습니다!
몇가지 추가 답변 드렸습니다 확인부탁드려요!

`JPA`미션 3단계 `merge`하겠습니다.

남은 미션 또한 화이팅입니다!
